File: cs/github.py
"""GitHub REST API via a per-identity fine-grained token (no gh CLI needed).

Tokens are keyed by GitHub owner (user or org login) and live in
~/.config/claude-share/tokens/<owner> (0600, never synced) or the env var CS_GITHUB_TOKEN_<OWNER>. Required token permissions on the owner:
Repository administration (read/write) and Metadata (read); for an organization
owner the token must be authorized for that org.
"""
from __future__ import annotations

import getpass
import json
import os
import urllib.error
import urllib.request
from pathlib import Path
from typing import Any, Dict, Optional

from . import paths

API = "https://api.github.com"


class GitHubError(RuntimeError):
    pass


def token_file(owner: str) -> Path:
    return paths.cs_config_dir() / "tokens" / owner.lower()


def get_token(owner: str) -> Optional[str]:
    env = os.environ.get(f"CS_GITHUB_TOKEN_{owner.upper().replace('-', '_')}")
    if env:
        return env.strip()
    f = token_file(owner)
    if f.exists():
        return f.read_text().strip() or None
    return None


def set_token(owner: str, token: Optional[str] = None) -> Path:
    if not token:
        token = getpass.getpass(f"GitHub fine-grained token for '{owner}' (hidden; needs Administration r/w on all repos): ").strip()
    if not token:
        raise SystemExit("cs: empty token")
    f = token_file(owner)
    f.parent.mkdir(parents=True, exist_ok=True)
    f.parent.chmod(0o700)
    f.write_text(token + "\n")
    f.chmod(0o600)
    return f


def api(method: str, path: str, token: str, body: Optional[Dict[str, Any]] = None) -> Any:
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(API + path, data=data, method=method, headers={
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
        "User-Agent": "claude-share",
        **({"Content-Type": "application/json"} if data else {}),
    })
    try:
        with urllib.request.urlopen(req, timeout=20) as r:
            raw = r.read()
            return json.loads(raw) if raw else {}
    except urllib.error.HTTPError as e:
        try:
            msg = json.loads(e.read()).get("message", "")
        except Exception:
            msg = ""
        raise GitHubError(f"GitHub {method} {path}: {e.code} {msg}".strip())
    except urllib.error.URLError as e:
        raise GitHubError(f"GitHub unreachable: {e.reason}")


def whoami(token: str) -> str:
    return api("GET", "/user", token)["login"]


def owner_type(owner: str, token: str) -> str:
    """'User' or 'Organization'."""
    return api("GET", f"/users/{owner}", token)["type"]


def repo_exists(owner: str, name: str, token: str) -> bool:
    try:
        api("GET", f"/repos/{owner}/{name}", token)
        return True
    except GitHubError as e:
        if " 404 " in str(e) or str(e).endswith("404"):
            return False
        raise


def create_repo(owner: str, name: str, token: str, private: bool = True, description: str = "") -> Dict[str, Any]:
    body = {"name": name, "private": private, "description": description, "auto_init": False}
    if owner_type(owner, token) == "Organization":
        return api("POST", f"/orgs/{owner}/repos", token, body)
    me = whoami(token)
    if me.lower() != owner.lower():
        raise GitHubError(f"token belongs to '{me}', cannot create repos for user '{owner}'")
    return api("POST", "/user/repos", token, body)


def ensure_token(owner: str, interactive: bool = True) -> str:
    """Return a working token for owner, prompting for one if missing/invalid."""
    tok = get_token(owner)
    if tok:
        return tok
    if not interactive:
        raise GitHubError(f"no GitHub token for '{owner}' — run `cs token set {owner}`")
    set_token(owner)
    tok = get_token(owner) or ""
    me = whoami(tok)
    if owner_type(owner, tok) == "User" and me.lower() != owner.lower():
        raise GitHubError(f"token authenticates as '{me}', not '{owner}'")
    return tok


def ensure_repo(owner: str, name: str, token: str, private: bool = True, description: str = "") -> bool:
    """Create owner/name if missing. Returns True when it was created."""
    if repo_exists(owner, name, token):
        return False
    create_repo(owner, name, token, private=private, description=description)
    return True

File: src/github.ts
/** GitHub REST API with a per-owner fine-grained token (no gh CLI). Tokens: ~/.config/claude-share/tokens/<owner>. */
import { chmodSync, existsSync, mkdirSync, readFileSync, unlinkSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { csConfigDir } from "./paths.js";
import * as ui from "./ui.js";

export class GitHubError extends Error {}
export const tokenFile = (owner: string) => join(csConfigDir(), "tokens", owner.toLowerCase());
export function getToken(owner: string): string | undefined {
  const env = process.env[`CS_GITHUB_TOKEN_${owner.toUpperCase().replace(/-/g, "_")}`];
  if (env) return env.trim();
  const f = tokenFile(owner);
  return existsSync(f) ? readFileSync(f, "utf8").trim() || undefined : undefined;
}
export async function setToken(owner: string, token?: string) {
  token ||= await ui.password(`GitHub fine-grained token for '${owner}' (Administration r/w on all repos)`);
  if (!token) throw new Error("cs: empty token");
  const f = tokenFile(owner);
  mkdirSync(join(csConfigDir(), "tokens"), { recursive: true, mode: 0o700 });
  writeFileSync(f, token + "\n"); chmodSync(f, 0o600);
  return f;
}
export function rmToken(owner: string) { const f = tokenFile(owner); if (existsSync(f)) unlinkSync(f); }
export async function api<T = any>(method: string, path: string, token: string, body?: unknown): Promise<T> {
  let r: Response;
  try {
    r = await fetch("https://api.github.com" + path, { method, headers: { Authorization: `Bearer ${token}`, Accept: "application/vnd.github+json",
      "X-GitHub-Api-Version": "2022-11-28", "User-Agent": "claude-share", ...(body ? { "Content-Type": "application/json" } : {}) },
      body: body ? JSON.stringify(body) : undefined, signal: AbortSignal.timeout(20000) });
  } catch (e: any) { throw new GitHubError(`GitHub unreachable: ${e?.message ?? e}`); }
  const text = await r.text();
  if (!r.ok) { let msg = ""; try { msg = JSON.parse(text).message ?? ""; } catch {} throw new GitHubError(`GitHub ${method} ${path}: ${r.status} ${msg}`.trim()); }
  return text ? JSON.parse(text) : ({} as T);
}
export const whoami = async (t: string) => (await api("GET", "/user", t)).login as string;
export const ownerType = async (o: string, t: string) => (await api("GET", `/users/${o}`, t)).type as "User" | "Organization";
export async function repoExists(o: string, n: string, t: string) {
  try { await api("GET", `/repos/${o}/${n}`, t); return true; } catch (e) { if (String(e).includes(" 404")) return false; throw e; }
}
export async function createRepo(o: string, n: string, t: string, priv = true, description = "") {
  const body = { name: n, private: priv, description, auto_init: false };
  if ((await ownerType(o, t)) === "Organization") return api("POST", `/orgs/${o}/repos`, t, body);
  const me = await whoami(t);
  if (me.toLowerCase() !== o.toLowerCase()) throw new GitHubError(`token belongs to '${me}', cannot create repos for user '${o}'`);
  return api("POST", "/user/repos", t, body);
}
export async function ensureToken(owner: string, interactive = true): Promise<string> {
  const t = getToken(owner);
  if (t) return t;
  if (!interactive) throw new GitHubError(`no GitHub token for '${owner}' — run cs token set ${owner}`);
  await setToken(owner);
  const tok = getToken(owner)!;
  const me = await whoami(tok);
  if ((await ownerType(owner, tok)) === "User" && me.toLowerCase() !== owner.toLowerCase()) throw new GitHubError(`token authenticates as '${me}', not '${owner}'`);
  return tok;
}
export async function ensureRepo(o: string, n: string, t: string, priv = true, description = "") {
  if (await repoExists(o, n, t)) return false;
  await createRepo(o, n, t, priv, description);
  return true;
}
